consensus/istanbul/backend/announce.md

```diff
@@ -0,0 +1,217 @@
+- `timestamp`: a message generation timestamp, used to bypass hash caches for messages.
+
+Max amount of encrypted node urls is 2 * size(set of [NearlyElectedValidator]).
+
+### enodeCertificateMsg (0x17)
+
+This message holds only ONE [eNodeURL] and it's meant to be send directly from [NearlyElectedValidator] to [NearlyElectedValidator] in a direct
+connection as a response to a [queryEnodeMsg].
+
+`[enode_url: B, version: P]`
+
+- `enode_url`: the origin's plaintext `eNodeURL`.
+- `version`: the current announce version for the origin's `eNodeURL`.
+
+### versionCertificatesMsg (0x16)
+
+This messages holds MANY version certificates. It is used mostly in two ways:
+
+- To share the WHOLE version table a [FullNode] has.
+- To share updated parts of the table of a [FullNode].
+
+`[[version: P, signature: B]]`
+
+- `version`: the current highest known announce version for a validator (deduced from the signature), according to the peer that gossipped this message.
+- `signature`: the signature for the version payload string (`'versionCertificate'|version`) by the emmitter of the certificate. Note that the address and public key can be deduced from this `signature`.
+
+## Spec by node type
+
+Note that validators are full nodes too, therefore they should follow the same rules as full nodes. This is specially important to not give away the status of the node as a validator.
+
+### Full node (non validator) Spec
+
+#### Peer registered
+
+When a peer is registered, the whole version table should be sent to the peer in a [versionCertificatesMsg].
+
+#### Handling [queryEnodeMsg]
+
+Messages received of this type should be only processed once.
+Should be regossipped as is, unless another message from the same validator origin has been regossipped in the past 5 minutes.
+
+#### Handling [enodeCertificateMsg]
+
+[enodeCertificateMsg] should be ignored by [FullNode] instances, since they should never receive one if all other participants are behaving properly. We make
+this explicit in the spec since the regossipping of this particular message would make the [eNodeURL] of the sender public, which defeats the purpose of the
+protocol.
+
+#### Handling [versionCertificatesMsg]
+
+Messages received of this type should be only processed once.
+When received, it should upsert its version table, and then multicast to its peers a new filtered [versionCertificatesMsg] without the certificates that were not updated in the table, and also filtering out certificates if another one for the same [NearlyElectedValidator] has been gossiped in the past 5 minutes.
+
+#### Message spawning
+
+Every 5 minutes it should multicast its version certificate table to all peers.
+
+### Validator Spec
+
+#### Peer handshake
+
+When a [NearlyElectedValidator] connects to another [NearlyElectedValidator], the inbound peer can send an [enodeCertificateMsg] to identify itself as a [NearlyElectedValidator]. This allows for preferential treatment for the p2p connection.
```

Comment:
I actually prefer inbound as opposed to initiator since it then affords the use of outbound to describe the other side of the connection which is useful. Otherwise I guess you could use initiator and non-initiator but it seems a bit clunky. Also inbound and outbound are the terms currently used in the codebase to describe this property of connections.

Reply:
i guess, what confuses me is that inbound/output referes to a connection, not to the to the actor. The actor is either the initiator or the receiving part of it.